### DISCUSSION TITLE
ScrollView Not Working, Misunderstanding?

### DISCUSSION THREAD

**Discussion by dynamoRando** [tui-cs/Terminal.Gui#1049] [Q&A] [answered]
I've started a repo where I want to build a terminal version of something akin to Azure Data Studio or Microsoft SQL Server Management Studio.

I'm running this on Ubuntu 20.04 as a .net5.0 project / netstandard 2.1. For this demo I've been using the [AdventureWorksDW2019](https://docs.microsoft.com/en-us/sql/samples/adventureworks-install-configure?view=sql-server-ver15&tabs=ssms) database on MS SQL Server on my local machine. To compare results, I've been using Azure Data Studio.

I'm having trouble getting the ScrollView to work. My code is [here](https://github.com/dynamoRando/SQLine/blob/dev_ui/SQLineGUI/UI/ConsoleOutput.cs). I apologize in advance, this is all very rough proof of concept code and needs to be cleaned up.

In this class, I have a Window, which has a TextView and a ScrollView. The TextView is added to the ScrollView, and then ScrollView is added to the Window.

Basically, when the user executes a SQL query and gets the result set back, I return to the UI a List<string> of formatted strings representing the data set, to be rendered in the TextView.

I am expecting the ScrollView to automatically add scrollbars to the UI and to be able to scroll left, right, up, down for very large data sets. However, this almost never works.

I thought perhaps the problem was that I needed to dynamically adjust the ContentSize, so now I've tried calculating the max width and height that the TextView is going to display and setting that as the ContentSize for the ScrollView, but that doesn't seem to help either. 

Basically, sometimes the horizontal scrollbar works, and sometimes it doesn't, especially if I am re-sizing the console window itself. The vertical scrollbar almost never works.

Assuming you have the same environment as me, reproduction steps (in the branch **dev_ui**)  -

1. Start the application
2. In the "Console" window in my application, type "cn -sn [serverName] -un [UserName] -pw [password]" to set the connection string values
3. Type "use AdventureWorksDW2019" to set the database as AdventureWorksDW2019 (you may hit tab to autocomplete)
4. Type "q select * from dimdate" to query the DimDate table from the AdventureWorksDW2019 database

From here, observe that the "Output" window that I have attempts to render the result set, but the scrollbars don't appear to work at all, or sometimes occasionally works if you resize the OS's console window - at least for me, the horizontal scrollbar works, but the vertical one doesn't work period. I've been banging my head on this for about 5 days now, so any help is appreciated.

Any suggestions are welcome on improving my code / what to change / my mis-understanding of how the gui.cs framework is supposed to work.

Thank you for your time.

**Comment by tznind** [comment]
> Basically, when the user executes a SQL query and gets the result set back, I return to the UI a List of formatted strings representing the data set, to be rendered in the TextView.

I can't help with the scroll view but if you want to show a result set in a formatted table (with horizontal and vertical scrolling) then I would appreciate any feedback on my PR (TableView) since this is exactly what it is designed to do (show tabular data via `System.DataTable` in user friendly form): https://github.com/migueldeicaza/gui.cs/pull/1015 .  It should be pretty easy to use the new control even if you are consuming the nuget package directly by importing just the code in TableView.cs

**Reply by dynamoRando** [reply]
Just tested this AM, looks great! This should do what I need to do, thank you!

**Comment by migueldeicaza** [comment, marked as answer]
Hello,

Short version: you do not need the scrollview today.

TextView handles the content on its own, and does not have a use for the ScrollView.   I can see that would be useful, but at this point, if you would like to make this work, you would need to do a lot of work to essentially have the TextView Frame track the size of the contents, and then add that to the scrollview - and you would need to manage the cursor on your own.

Your concern is a valid one, it would be nice to support scrollbars on the TextView, please file an issue.

**Comment by BDisp** [comment]
See this PR #1081.
